### PR TITLE
[#32222] Actually maintain the heap invariant for timers.

### LINKED ISSUE #32222: [prism] Resolve endless timer test org.apache.beam.sdk.transforms.ParDoTest$TimerTests.testEventTimeTimerOrderingWithCreate
```
TEST=org.apache.beam.sdk.transforms.ParDoTest\$TimerTests.testEventTimeTimerOrderingWithCreate
./gradlew :runners:portability:java:ulrLoopbackValidatesRunnerTests -PjobEndpoint=localhost:8073 --tests="$TEST"
```

Doesn't currently terminate, which means there's a subtle ordering requirement that prism isn't respecting somehow, leading to infinite looping. 

The way the test expects to execute needs to be described here, so we can determine what's going wrong, and why it doesn't terminate.

### PR DESCRIPTION
**Pull request by lostluck**
testEventTimeTimerOrderingWithCreate was hanging due to timers being fired in incorrect order. The issue was that the heap variant wasn't being maintained on inserting timers for a key. This would cause timers to fire out of event time ordering. In particular, the "GC" timer would fire first (because it was set first), instead of the earlier firing "append" timer.

This incorrect ordering was also what caused testTwoTimersSettingEachOtherWithCreateAsInputUnbounded to fail.

Fixes #32222.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @damondouglas

**Comment by lostluck** [comment]
Fix two Java tests, but now a Python test is failing. Curious.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner_test.py [hunk 1]
```diff
@@ -790,8 +791,10 @@ def is_buffered_correctly(actual):
       # Acutal should be a grouping of the inputs into batches of size
       # at most buffer_size, but the actual batching is nondeterministic
       # based on ordering and trigger firing timing.
-      self.assertEqual(sorted(sum((list(b) for b in actual), [])), elements)
-      self.assertEqual(max(len(list(buffer)) for buffer in actual), buffer_size)
+      self.assertEqual(
```
Comment (lostluck) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner_test.py:794]: Added the actual output to the asserts help with debugging, hence the line reflows.

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go [hunk 2]
```diff
@@ -1159,7 +1159,7 @@ func (ss *stageState) AddPending(newPending []element) int {
 				}
 				ss.pendingByKeys[string(e.keyBytes)] = dnt
 			}
-			dnt.elements.Push(e)
+			heap.Push(&dnt.elements, e)
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go:1162]: This fixes maintaining the heap invariant.

**Review by robertwb — COMMENTED** [review]
Test fix LGTM.

**Review by damondouglas — APPROVED** [review]

**Comment by lostluck** [comment]
Thanks!

I've made a note to log a clear warning/error when timers are set beyond their window + AllowedLateness. Once that's in, it'll be clearer whether I can just hard fail pipelines when it happens with a hopefully clear error message.

**Pull request merged**